File: crates/luu/ui/workspace.js
// @ts-check
/// The workspace panels' shared state: which folder is being looked at, what
/// the tree has open, what git says, and what the content column is showing.
///
/// A module rather than props threaded through `app.html`, for the reason
/// `store.js` is one: the Files panel, the Git panel and the content viewer are
/// siblings, and a parent whose only job is relaying between its children is a
/// parent the state does not need. Both panels write tabs; the viewer reads
/// them. See `RECORD/2026-09-15.a-three-pane-inspector.WIP.md` and
/// `RECORD/2026-09-16.three-columns-that-each-have-a-footer.completed.md`.

import { $reactive } from "./vendor/jq79.js"
import { apiHeaders } from "./store.js"
import { setPane } from "./prefs.js"

export const workspace = $reactive({
  /// The absolute path `luu serve` was started in, as the server reports it.
  /// Display only — nothing is ever asked for by absolute path, because an
  /// absolute path in a URL is an invitation to send a different one.
  base: "",
  /// The chosen folder, relative to the base and `/`-separated. `""` is the
  /// base itself. **`null` means nobody has chosen yet**, which is what opens
  /// the picker over the whole window on a first visit — it is not the same
  /// state as `""`, and collapsing the two would mean the picker either never
  /// appears or appears every time.
  root: null,
  /// One entry per directory that has been opened, keyed by its path relative
  /// to the sandbox base — never to the root, so narrowing the root does not
  /// invalidate a listing that is still correct. `""` is the base itself.
  /// Lazy, like every file tree: a repository with a `target/` in it is not
  /// something to walk eagerly.
  open: {},
  /// Directories currently expanded, in the order they were opened. A plain
  /// object beside `open` rather than a flag inside it: collapsing a directory
  /// should not throw away what was already fetched.
  expanded: {},
  /// What the content column is holding: `{ id, kind, title, path, staged }`,
  /// where `kind` is `file`, `diff` or `debug`. One tab per open thing, so
  /// looking at a diff no longer loses the file that was on screen.
  tabs: [],
  /// The active tab's id, or `null` for an empty column.
  active: null,
  /// The active tab itself. `picked` rather than `selected` in the components
  /// that read it, and the name is load-bearing: a scope variable called
  /// `selected` collides with the DOM property of that name, and jq79's render
  /// effect ends up writing what it reads. See `content-viewer.html`.
  selected: null,
  /// What the viewer is showing, once it has arrived.
  /// `{ kind: "file", path, lines, truncated }`,
  /// `{ kind: "diff", path, staged, hunks, note }` or
  /// `{ kind: "debug", text }`.
  content: null,
  /// Set while a fetch for the active tab is in flight, so the viewer can say
  /// so instead of showing the previous file under the new file's name.
  loading: false,
  /// The file's rows, in the two blocks the viewer renders as two `:each`
  /// groups: `head` is the first screenful and `tail` is everything after it,
  /// one frame later.
  ///
  /// **Two arrays rather than one that grows.** The file arrives whole and
  /// already coloured; what costs is the row per line — ~50 µs, measured, with
  /// or without colour — and jq79 rebuilds an `:each` rather than appending to
  /// it. A growing array therefore rebuilds the rows already on screen every
  /// time it grows; two arrays mean the second render touches only the second
  /// block and the first is never built twice. See the phase 8 sections of
  /// `RECORD/2026-09-15.a-three-pane-inspector.WIP.md`.
  head: [],
  tail: [],
  /// How many rows `tail` *will* have, for the spacer that keeps the scrollbar
  /// the length of the file rather than the length of the first screen while
  /// the tail is still being built.
  pending: 0,
  /// Every changed path in the whole base, mapped to git's own two-letter code.
  /// Asked for the base and filtered to the root in the page rather than
  /// parameterised: it is one `git status --porcelain` for a repository either
  /// way, and a second parameter would be a second thing to keep consistent
  /// with the prefix.
  status: {},
  /// Why git could not be asked, when it could not. A workspace that is not a
  /// git repository is an ordinary case and lands here as a plain sentence.
  gitError: null,
  /// The last failure from any of these endpoints, for the panel to show.
  error: null,
  /// The icon theme's maps, by icon id, or `{ loaded: false }` when this
  /// machine named none. See `crate::icons` for why nothing is vendored, and
  /// `iconFor` for why the tree asks it per row rather than reading this flag.
  icons: { loaded: false },
  /// The folder picker: whether it is open, and the directory it is showing.
  /// `forced` is a first visit, where it covers the window and has no way out
  /// other than choosing — there is nothing behind it to go back to.
  picker: { open: false, forced: false, at: "" },
})

// ---- which folder ---------------------------------------------------------
//
// The folder `luu serve` was started in is the **ceiling**, and what is chosen
// is a subdirectory of it. Nothing here widens anything: `/api/workspace/*`
// resolves every path through `Sandbox::check_path` against that base, and a
// chosen root is a prefix this page puts on what it asks for. A picker that
// could climb above the base would be a browser control that moves the sandbox,
// and that is the one thing the workspace surface has never been.

/// Where the remembered root is kept, keyed by the base — two servers on one
/// machine do not inherit each other's answer.
const rootKey = base => `luu.root:${base}`

/// A base-relative path as the panels *show* it — relative to the chosen root.
/// Returns `null` for anything outside it, which is what filters git's status
/// down to the folder somebody chose.
///
/// **Only the display goes through here.** Every path this module sends to the
/// server stays relative to the base, exactly as the server hands it out: a
/// second spelling for the same file is a second chance to get the prefix wrong,
/// and the root is a narrowing of what is *listed*, not a second address space.
export function inside(path) {
  const root = workspace.root || ""
  if (!root) return path
  if (path === root) return ""
  return path.startsWith(`${root}/`) ? path.slice(root.length + 1) : null
}

/// Reads the base off the settings the server already answers, and the root off
/// what this browser remembered. Opens the picker when nothing was.
export async function loadRoot(base) {
  workspace.base = base || ""
  let remembered = null
  try {
    remembered = localStorage.getItem(rootKey(workspace.base))
  } catch {
    // A private window. The picker opens; the choice is simply not kept.
  }
  if (remembered === null) {
    workspace.picker = { open: true, forced: true, at: "" }
    return false
  }
  workspace.root = remembered
  return true
}

export function openPicker() {
  workspace.picker = { open: true, forced: false, at: workspace.root || "" }
  // The picker lists directories, which means listing them: a folder nobody has
  // opened in the tree has never been fetched.
  openDir(workspace.picker.at)
}

export function closePicker() {
  // A first visit has nothing behind it to go back to, so the only way out of a
  // forced picker is choosing.
  if (workspace.picker.forced) return
  workspace.picker = { open: false, forced: false, at: "" }
}

export function browseTo(path) {
  workspace.picker = { ...workspace.picker, at: path }
  openDir(path)
}

/// Takes the folder, and starts the panels over on it. Everything that was open
/// was open relative to a different root, so none of it survives: a tab holding
/// a file outside the new root would be a tab the tree cannot reach.
export async function chooseRoot(path) {
  workspace.root = path
  try {
    localStorage.setItem(rootKey(workspace.base), path)
  } catch {
    // Applied for this visit, and asked again next time.
  }
  workspace.picker = { open: false, forced: false, at: "" }
  workspace.tabs = []
  workspace.active = null
  workspace.selected = null
  workspace.content = null
  workspace.head = []
  workspace.tail = []
  workspace.pending = 0
  workspace.open = {}
  workspace.expanded = {}
  await openDir(path, true)
  await loadStatus()
}

/// The icon id for one entry, by VSCode's own rules.
///
/// Full filename first, then progressively shorter extensions — so `d.ts`
/// beats `ts` for `index.d.ts`, which is the distinction those themes draw —
/// then the theme's default. `null` when this row has no icon in this theme —
/// no theme loaded, or a theme that declares no default — and the tree draws
/// its own file or folder instead.
export function iconFor(name, isDir, expanded = false) {
  const theme = workspace.icons
  if (!theme.loaded) return null
  const lower = name.toLowerCase()
  if (isDir) {
    return (
      theme.folder_names?.[lower] ??
      (expanded ? theme.folder_expanded : theme.folder) ??
      theme.folder ??
      null
    )
  }
  const named = theme.file_names?.[lower]
  if (named) return named
  const parts = lower.split(".")
  for (let at = 1; at < parts.length; at++) {
    const suffix = parts.slice(at).join(".")
    const found = theme.file_extensions?.[suffix]
    if (found) return found
  }
  return theme.file ?? null
}

export async function loadIcons() {
  try {
    workspace.icons = await ask("./api/icons/manifest")
  } catch {
    // A page that cannot read the manifest simply has no icons; the tree is
    // the point and it still lists.
    workspace.icons = { loaded: false }
  }
}

async function ask(url) {
  const answer = await fetch(url, { headers: apiHeaders() })
  if (!answer.ok) throw new Error((await answer.text()) || `HTTP ${answer.status}`)
  return answer.json()
}

/// Lists one directory, caching it. `force` re-reads one already listed, which
/// is what the refresh button wants. `path` is relative to the base.
export async function openDir(path = "", force = false) {
  const key = path || ""
  if (!force && workspace.open[key]) {
    workspace.expanded = { ...workspace.expanded, [key]: true }
    return
  }
  try {
    const tree = await ask(`./api/workspace/tree?path=${encodeURIComponent(key)}`)
    workspace.open = { ...workspace.open, [key]: tree.entries }
    workspace.expanded = { ...workspace.expanded, [key]: true }
    workspace.gitError = tree.git || null
    workspace.error = null
  } catch (e) {
    workspace.error = `Could not list ${key || "the workspace"}: ${e.message}`
  }
}

export function closeDir(path) {
  const next = { ...workspace.expanded }
  delete next[path]
  workspace.expanded = next
}

export function toggleDir(path) {
  if (workspace.expanded[path]) closeDir(path)
  else openDir(path)
}

/// Re-reads every directory that is currently open, and git with it. What the
/// refresh button does: a tree is a snapshot, and an agent that just wrote a
/// file has made it stale.
export async function refresh() {
  const keys = Object.keys(workspace.open)
  await Promise.all(keys.map(key => openDir(key, true)))
  await loadStatus()
}

export async function loadStatus() {
  try {
    workspace.status = await ask("./api/workspace/git-status")
    workspace.gitError = null
  } catch (e) {
    workspace.status = {}
    workspace.gitError = e.message
  }
}

// ---- the content column's tabs --------------------------------------------
//
// **A tab holds its identity, not its bytes.** Activating one re-fetches. The
// server answers the largest file in this repository in 20 ms and a small one in
// 0.6; the 271 ms a file costs is building the rows, which a cache would not
// save because the rows are rebuilt on activation either way. So a cache would
// hold megabytes of JSON to save 20 ms of the 291 that switching a tab costs.
// See `RECORD/2026-09-16.what-the-debug-ui-does-not-need.completed.md`.

const fileId = path => `file:${path}`
// **One tab per file's diff, not one per side.** "Staged" and "working tree"
// are two different answers about the same file, and the foot's toggle is how
// you read the other one — a second tab with the same name in it would make the
// strip say there are two files open when there is one.
const diffId = path => `diff:${path}`

const basename = path => path.split("/").pop() || path

/// Opens a tab, or focuses the one already holding this thing, and points the
/// second column at the content — the click already said which column the
/// person wants, and in the two-column layout having nothing happen is the bug
/// this rule exists to prevent.
function openTab(tab) {
  const found = workspace.tabs.find(open => open.id === tab.id)
  // Replaced rather than mutated, the way every reactive list in this page is:
  // jq79 does not wake an `:each` when a property of an object inside a
  // reactive array is assigned from outside the component. See `store.js`.
  workspace.tabs = found
    ? workspace.tabs.map(open => open.id === tab.id ? { ...open, ...tab } : open)
    : [...workspace.tabs, tab]
  setPane("content")
  return activate(tab.id)
}

export function closeTab(id) {
  const at = workspace.tabs.findIndex(tab => tab.id === id)
  if (at < 0) return
  const left = workspace.tabs.filter(tab => tab.id !== id)
  workspace.tabs = left
  if (workspace.active !== id) return
  // Falls to the neighbour on the right, then on the left: closing the tab you
  // are looking at should land you next to it rather than at the far end.
  const next = left[at] || left[at - 1] || null
  if (next) {
    activate(next.id)
  } else {
    workspace.active = null
    workspace.selected = null
    workspace.content = null
    workspace.head = []
    workspace.tail = []
    workspace.pending = 0
  }
}

export function activate(id) {
  const tab = workspace.tabs.find(open => open.id === id)
  if (!tab) return
  workspace.active = id
  // A **copy**, never the element out of `tabs`. Putting the same object at two
  // paths in a reactive store means every read of `selected` re-wraps what is
  // already wrapped under `tabs`, the identity changes on every pass, and jq79
  // gives up after 100 of them with "an effect re-woke itself" — which names
  // neither this line nor this file. Copying is what makes the two paths two
  // values. See `RECORD/2026-09-16.three-columns-that-each-have-a-footer.completed.md`.
  workspace.selected = { ...tab }
  if (tab.kind === "file") return loadFile(tab)
  if (tab.kind === "diff") return loadDiff(tab)
  // A debug tab carries its own text: there is nothing to fetch, because the
  // context panel already had it.
  workspace.content = { kind: "debug", text: tab.text, path: tab.title }
  workspace.head = []
  workspace.tail = []
  workspace.pending = 0
  workspace.loading = false
}

/// Rows in the first pass: a screenful with room to scroll into, and nothing
/// to do with the window's real height — a viewer that measures itself has to
/// decide again on every resize.
const FIRST_ROWS = 200

/// Fills the second block, one frame after the first one is on screen.
function showTheRest(id, lines) {
  if (lines.length <= FIRST_ROWS) return
  requestAnimationFrame(() => {
    // Dropped if the viewer moved on while this frame was waiting: these rows
    // belong to the tab on screen, never to the one that was.
    if (workspace.active !== id) return
    workspace.tail = lines.slice(FIRST_ROWS)
    workspace.pending = 0
  })
}

export function showFile(path) {
  return openTab({ id: fileId(path), kind: "file", path, title: basename(path), staged: false })
}

export function showDiff(path, staged = false) {
  return openTab({ id: diffId(path), kind: "diff", path, staged, title: basename(path) })
}

/// A prompt, a trace, a tool's output: text the context panel already has, in
/// the column that is wide enough to read it. The one thing a 20rem rail has
/// never been able to show.
export function showDebug(id, title, text) {
  return openTab({ id: `debug:${id}`, kind: "debug", path: title, title, text })
}

async function loadFile(tab) {
  workspace.loading = true
  try {
    const file = await ask(`./api/workspace/file?path=${encodeURIComponent(tab.path)}`)
    // Dropped if the person clicked something else while this was in flight:
    // the column's tab and its body have to be the same file.
    if (workspace.active !== tab.id) return
    workspace.content = { kind: "file", ...file }
    workspace.head = file.lines.slice(0, FIRST_ROWS)
    workspace.tail = []
    workspace.pending = Math.max(0, file.lines.length - FIRST_ROWS)
    showTheRest(tab.id, file.lines)
    workspace.error = null
  } catch (e) {
    workspace.content = null
    workspace.error = `Could not read ${tab.path}: ${e.message}`
  } finally {
    if (workspace.active === tab.id) workspace.loading = false
  }
}

async function loadDiff(tab) {
  workspace.loading = true
  try {
    const diff = await ask(
      `./api/workspace/git-diff?path=${encodeURIComponent(tab.path)}&staged=${tab.staged}`,
    )
    if (workspace.active !== tab.id) return
    workspace.content = { kind: "diff", ...diff }
    workspace.head = []
    workspace.tail = []
    workspace.pending = 0
    workspace.error = null
  } catch (e) {
    workspace.content = null
    workspace.error = `Could not diff ${tab.path}: ${e.message}`
  } finally {
    if (workspace.active === tab.id) workspace.loading = false
  }
}

File: crates/luu/src/workspace.rs
//! The workspace, as the person at the browser looks at it: a file tree, what
//! git says changed, and the contents of one file or one diff.
//!
//! **Read-only, and outside the job gate on purpose.** Every other way this
//! program touches the filesystem is a *model's* action — `read_file` and
//! `run_command` are tools, and a tool call is held until somebody approves the
//! job that proposed it. Nothing here is proposed by anything: a person clicked
//! a directory. Routing that through the approval flow would mean either
//! rubber-stamping a job nobody asked the model to make, or building a second
//! read path that skips the gate — and then there are two ways to read a file
//! and only one of them is audited. The read side of the session API
//! (`/api/sessions/*`) already answers plain ungated `GET`s for the same
//! reason. See `RECORD/2026-09-15.a-three-pane-inspector.WIP.md`.
//!
//! What still bounds it is the sandbox: every path is resolved through
//! [`Sandbox::check_path`], so this surface can reach exactly what the session's
//! own policy grants and nothing above it.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use std::process::Stdio;

use agent_core::sandbox::{Access, Sandbox};
use serde::Serialize;
use tokio::process::Command;

/// How long any one `git` call may take before it is abandoned. A repository
/// on a dead network mount is the case this exists for: the panel showing
/// "could not read git" beats the page hanging on a click.
const GIT_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);

/// One entry in a directory listing.
#[derive(Debug, Clone, Serialize)]
pub struct Entry {
    /// Relative to the sandbox base, `/`-separated, and the id the page sends
    /// back to open it. Never absolute: an absolute path in a URL is an
    /// invitation to send a different one.
    pub path: String,
    pub name: String,
    pub dir: bool,
    /// Whether git ignores it. Shown dimmed rather than hidden: a person
    /// looking for `target/` wants to see that it is there and ignored, not to
    /// wonder whether the tree is broken.
    pub ignored: bool,
    /// Git's own two-letter status, when it has one — `M`, `??`, `A`, `D`.
    /// `None` is unchanged, which is most of a tree.
    pub status: Option<String>,
}

/// One directory's children.
#[derive(Debug, Serialize)]
pub struct Tree {
    /// What was listed, relative to the base. Empty string is the base itself.
    pub path: String,
    pub entries: Vec<Entry>,
    /// What went wrong asking git, when something did. The tree still lists —
    /// a workspace that is not a git repository at all is an ordinary case,
    /// not an error, and it renders as a tree with nothing marked.
    pub git: Option<String>,
}

/// A file, read for display.
#[derive(Debug, Serialize)]
pub struct FileView {
    pub path: String,
    pub text: String,
    /// True when the file was cut at [`MAX_FILE_BYTES`]. The page says so
    /// rather than showing a truncated file as if it were whole.
    pub truncated: bool,
}

/// The cap on a file the viewer will render. A debug UI has no business
/// streaming a gigabyte into a browser, and the number is the same order as
/// the tool output cap the agent already lives under.
pub const MAX_FILE_BYTES: usize = 512 * 1024;

/// One line of a diff, already classified so the browser renders structure
/// rather than parsing text.
#[derive(Debug, Serialize)]
pub struct DiffLine {
    /// `add`, `del`, or `ctx`.
    pub kind: &'static str,
    pub text: String,
}

#[derive(Debug, Serialize)]
pub struct Hunk {
    /// Git's own `@@ … @@` header, kept verbatim: it is the one line here a
    /// person may want to read as git wrote it.
    pub header: String,
    pub lines: Vec<DiffLine>,
}

#[derive(Debug, Serialize)]
pub struct Diff {
    pub path: String,
    /// Staged against `HEAD` when `staged`, working tree against the index
    /// otherwise.
    pub staged: bool,
    pub hunks: Vec<Hunk>,
    /// Set when git answered but there was nothing to show — a file that is
    /// untracked has no diff, and saying so beats an empty panel.
    pub note: Option<String>,
}

/// Why a request could not be answered. Kept as one type because every one of
/// these becomes the same shape of HTTP error, and the caller should not have
/// to remember which variant maps to which status.
#[derive(Debug)]
pub enum Error {
    /// The path escaped the sandbox, or the sandbox does not grant it.
    Refused(String),
    /// It is not there.
    Missing(String),
    /// Something under it failed — a read, a spawn, a decode.
    Failed(String),
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Error::Refused(why) | Error::Missing(why) | Error::Failed(why) => write!(f, "{why}"),
        }
    }
}

/// Resolves a caller-supplied relative path against the base, through the
/// sandbox.
///
/// The sandbox is what makes this safe rather than any string handling here:
/// `check_path` resolves symlinks and `..` and answers against the policy's
/// roots, which is the same check `read_file` goes through.
fn resolve(sandbox: &Sandbox, relative: &str) -> Result<PathBuf, Error> {
    let relative = relative.trim_start_matches('/');
    let joined = match relative.is_empty() {
        true => sandbox.base().to_path_buf(),
        false => sandbox.base().join(relative),
    };
    let check = sandbox.check_path(&joined, Access::Read);
    if !check.verdict.allowed {
        return Err(Error::Refused(format!(
            "{relative}: {}",
            check.verdict.rule
        )));
    }
    Ok(check.path)
}

/// The path an entry is known by: relative to the base, `/`-separated.
fn relative_of(base: &Path, path: &Path) -> String {
    path.strip_prefix(base)
        .unwrap_or(path)
        .components()
        .map(|part| part.as_os_str().to_string_lossy().into_owned())
        .collect::<Vec<_>>()
        .join("/")
}

/// One directory's children, with git's opinion of each attached.
///
/// Ignored entries are found by difference rather than by a second matcher:
/// `ignore`'s walker answers "what is *not* ignored" for this directory, and
/// `read_dir` answers "what is there". Everything in the second and not the
/// first is ignored — which gets nested `.gitignore` files, negations and the
/// global config right for free, because the walker already implements all of
/// that and a hand-rolled matcher on the root file alone would not.
pub async fn tree(sandbox: &Sandbox, relative: &str) -> Result<Tree, Error> {
    let dir = resolve(sandbox, relative)?;
    if !dir.is_dir() {
        return Err(Error::Missing(format!("{relative}: not a directory")));
    }

    let mut listed: Vec<(String, bool, PathBuf)> = Vec::new();
    let mut reader = tokio::fs::read_dir(&dir)
        .await
        .map_err(|error| Error::Failed(format!("{relative}: {error}")))?;
    while let Some(item) = reader
        .next_entry()
        .await
        .map_err(|error| Error::Failed(format!("{relative}: {error}")))?
    {
        let name = item.file_name().to_string_lossy().into_owned();
        // `.git` is machinery, not content. It is the one directory whose
        // contents nobody opens this panel to read, and it is enormous.
        if name == ".git" {
            continue;
        }
        let is_dir = item
            .file_type()
            .await
            .map(|kind| kind.is_dir())
            .unwrap_or(false);
        listed.push((name, is_dir, item.path()));
    }

    let unignored = not_ignored(&dir);
    // A workspace that is not a git repository still has a tree worth listing,
    // so git failing is a note on the answer rather than an error instead of
    // one: nothing is marked, and the panel says why.
    let (statuses, git_error) = match git_status(sandbox.base()).await {
        Ok(map) => (map, None),
        Err(error) => (BTreeMap::new(), Some(error.to_string())),
    };

    let mut entries: Vec<Entry> = listed
        .into_iter()
        .map(|(name, dir_flag, full)| {
            let path = relative_of(sandbox.base(), &full);
            // A directory carries the loudest status under it, so a collapsed
            // tree still shows that something inside changed.
            let status = match dir_flag {
                false => statuses.get(&path).cloned(),
                true => {
                    let prefix = format!("{path}/");
                    statuses
                        .iter()
                        .find(|(key, _)| key.starts_with(&prefix))
                        .map(|(_, value)| value.clone())
                }
            };
            Entry {
                ignored: !unignored.contains(&name),
                path,
                name,
                dir: dir_flag,
                status,
            }
        })
        .collect();

    // Directories first, then by name: the order a file tree is read in.
    entries.sort_by(|a, b| b.dir.cmp(&a.dir).then_with(|| a.name.cmp(&b.name)));

    Ok(Tree {
        path: relative.trim_start_matches('/').to_string(),
        entries,
        git: git_error,
    })
}

/// The names in `dir` that git does *not* ignore.
///
/// Depth 1 and `parents(true)`: the walker has to read `.gitignore` files above
/// this directory to answer correctly for a subdirectory, and must not descend
/// below it.
fn not_ignored(dir: &Path) -> std::collections::HashSet<String> {
    ignore::WalkBuilder::new(dir)
        .max_depth(Some(1))
        .parents(true)
        .hidden(false)
        .git_ignore(true)
        .git_global(true)
        .git_exclude(true)
        .build()
        .filter_map(Result::ok)
        // The walker yields the root itself first; it is not one of its own
        // children.
        .filter(|entry| entry.path() != dir)
        .map(|entry| entry.file_name().to_string_lossy().into_owned())
        .collect()
}

/// Every path git considers changed, mapped to its two-letter status.
///
/// `-z` because a filename may contain anything, newlines and quotes included,
/// and `--porcelain`'s default output escapes those in a format that then has
/// to be un-escaped. The NUL-separated form is the one that needs no parsing
/// of quoting rules.
pub async fn git_status(base: &Path) -> Result<BTreeMap<String, String>, Error> {
    let out = git(base, &["status", "--porcelain=v1", "-z"]).await?;
    let mut map = BTreeMap::new();
    let mut fields = out.split('\0').filter(|part| !part.is_empty());
    while let Some(record) = fields.next() {
        // `XY <path>`, and for a rename `XY <to>\0<from>` — the extra field is
        // consumed here so it is not read as the next record.
        //
        // **Both characters, untrimmed.** X is the index against `HEAD` and Y
        // the working tree against the index, so ` M` and `M ` are opposite
        // facts and the space carries which one this is. Trimming it made
        // every unstaged change look staged, and the diff panel asked git for
        // the staged side and got nothing.
        let (code, path) = match record.len() > 3 {
            true => (record[..2].to_string(), record[3..].to_string()),
            false => continue,
        };
        if code.starts_with('R') || code.starts_with('C') {
            let _ = fields.next();
        }
        map.insert(path, code);
    }
    Ok(map)
}

/// A file, for the middle column.
pub async fn file(sandbox: &Sandbox, relative: &str) -> Result<FileView, Error> {
    let path = resolve(sandbox, relative)?;
    if path.is_dir() {
        return Err(Error::Missing(format!("{relative}: is a directory")));
    }
    let bytes = tokio::fs::read(&path)
        .await
        .map_err(|error| match error.kind() {
            std::io::ErrorKind::NotFound => Error::Missing(format!("{relative}: not found")),
            _ => Error::Failed(format!("{relative}: {error}")),
        })?;
    let truncated = bytes.len() > MAX_FILE_BYTES;
    let bytes = match truncated {
        true => &bytes[..MAX_FILE_BYTES],
        false => &bytes[..],
    };
    // Lossy rather than refused: a file with one bad byte in it is still worth
    // looking at, and the alternative is a panel that says nothing about a file
    // the tree says exists.
    Ok(FileView {
        path: relative.trim_start_matches('/').to_string(),
        text: String::from_utf8_lossy(bytes).into_owned(),
        truncated,
    })
}

/// One file's diff, parsed into hunks.
///
/// Parsed here rather than shipped as text, which is the call
/// `luu-design.md` already made for prompt diffs: "compute the diff in Rust and
/// send resolved spans". Git already produces a unified diff, so this parses
/// that rather than re-diffing the blobs — same destination, no second diff
/// implementation and no new dependency.
pub async fn diff(sandbox: &Sandbox, relative: &str, staged: bool) -> Result<Diff, Error> {
    // Resolved for the same reason a read is: a path that the sandbox refuses
    // must not become a `git diff` argument either.
    resolve(sandbox, relative)?;
    let path = relative.trim_start_matches('/');
    let mut args = vec!["diff", "--no-color", "--no-ext-diff"];
    if staged {
        args.push("--cached");
    }
    args.push("--");
    args.push(path);
    let out = git(sandbox.base(), &args).await?;

    let mut hunks: Vec<Hunk> = Vec::new();
    for line in out.lines() {
        if line.starts_with("@@") {
            hunks.push(Hunk {
                header: line.to_string(),
                lines: Vec::new(),
            });
            continue;
        }
        let Some(hunk) = hunks.last_mut() else {
            // Everything before the first `@@` is the file header git writes
            // (`diff --git`, `index`, `---`, `+++`), which the page already
            // knows: it asked about this path.
            continue;
        };
        let (kind, text) = match line.as_bytes().first() {
            Some(b'+') => ("add", &line[1..]),
            Some(b'-') => ("del", &line[1..]),
            Some(b'\\') => continue, // "\ No newline at end of file"
            _ => ("ctx", line.get(1..).unwrap_or("")),
        };
        hunk.lines.push(DiffLine {
            kind,
            text: text.to_string(),
        });
    }

    let note = match hunks.is_empty() {
        true => Some(match staged {
            true => "Nothing staged for this file.".to_string(),
            false => {
                "No unstaged changes — the file may be untracked, or already staged.".to_string()
            }
        }),
        false => None,
    };

    Ok(Diff {
        path: path.to_string(),
        staged,
        hunks,
        note,
    })
}

/// Runs git in the workspace and hands back its stdout.
///
/// **Not through the tool sandbox**, and that is the one thing worth being
/// explicit about: this is the server reading its own working directory on
/// behalf of the person looking at it, not a model running a command. It is
/// also why the `/dev/null` and `~/.gitconfig` findings in
/// `RECORD/2026-09-15.git-could-not-open-dev-null.completed.md` do not reach
/// it — a child here inherits the server's own environment.
async fn git(base: &Path, args: &[&str]) -> Result<String, Error> {
    let run = Command::new("git")
        .arg("-C")
        .arg(base)
        .args(args)
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true)
        .output();

    let out = tokio::time::timeout(GIT_TIMEOUT, run)
        .await
        .map_err(|_| Error::Failed("git did not answer in ten seconds".to_string()))?
        .map_err(|error| Error::Failed(format!("could not run git: {error}")))?;

    if !out.status.success() {
        let why = String::from_utf8_lossy(&out.stderr);
        let why = why.trim();
        return Err(Error::Failed(match why.is_empty() {
            true => format!("git exited with {}", out.status),
            false => why.to_string(),
        }));
    }
    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The rename record eats its second field. Written because the loop that
    /// does it is the one piece of parsing here that a reader cannot check by
    /// eye: `-z` output has no line breaks to count.
    #[test]
    fn a_rename_does_not_leave_its_old_name_as_a_record() {
        // `R  new\0old\0 M other\0` — one rename and one modification.
        let raw = "R  new\0old\0 M other\0";
        let mut map = BTreeMap::new();
        let mut fields = raw.split('\0').filter(|part| !part.is_empty());
        while let Some(record) = fields.next() {
            let (code, path) = match record.len() > 3 {
                true => (record[..2].to_string(), record[3..].to_string()),
                false => continue,
            };
            if code.starts_with('R') || code.starts_with('C') {
                let _ = fields.next();
            }
            map.insert(path, code);
        }
        // Both positions kept: "R " is a rename staged in the index, where
        // " R" would be one in the working tree, and the space is the only
        // thing that says which.
        assert_eq!(map.get("new").map(String::as_str), Some("R "));
        assert_eq!(map.get("other").map(String::as_str), Some(" M"));
        assert!(
            !map.contains_key("old"),
            "the old name is not its own entry"
        );
    }
}
